[Lai Wei Zhong] iP

Ken-Lai opened the pull request:
## Tasket

>“Drive thy business or it will drive thee.” – Benjamin Franklin ([source](https://dansilvestre.com/productivity-quotes/))

Tasket frees your mind of having to remember things you have to do, it's

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is
1. download it from [here](https://github.com/Ken-Lai/ip/releases)
2. double-click it
3. add your tasks
4. let it manage your task for you :wink:

And it's **FREE**!

Features:
- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

##
If you are a Java programmer, you can use it to learn Java too. Here's the main method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(Main.class, args)
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Ken-Lai committed:
Change chatbot name and implement an initial skeletal version

Ken-Lai committed:
Echoes command from user and exits when bye is entered

Ken-Lai committed:
Add ability to store text entered by user and display them back when
requested

Ken-Lai committed:
Add a Task class to represent tasks

Ken-Lai committed:
Add the ability to mark tasks as done or undone

Ken-Lai committed:
Implement ToDo, Event and Deadline to support multiple task types

Ken-Lai committed:
Add support to tracking three types of tasks: Todo, Deadline and
Event

Ken-Lai committed:
Semi-automate testing of text UIs with I/O redirection technique

Ken-Lai committed:
Add DukeException to handle errors specific to Duke

Ken-Lai committed:
Implement ability to handle invalid inputs entered by user

Ken-Lai committed:
Add support to delete task from the list

Ken-Lai committed:
Add ability to save and retrieve tasks from hard drive

Ken-Lai committed:
Merge branch 'branch-Level-7'

Ken-Lai committed:
Add ability to detect and convert date format in deadlines

Ken-Lai committed:
Merge branch 'branch-Level-8'

jiakai-17 commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,18 @@
+public class Deadline extends Task{
```
'{' is not preceded with whitespace. (1:35) [WhitespaceAround]

jiakai-17 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,358 @@
+import java.util.Scanner;
+
 public class Duke {
+    private static final String SEPARATION_LINE  = "____________________________________________________________";
```
Use a single space to separate non-whitespace characters. (12:50) [SingleSpaceSeparator]
Perhaps you could remove the extra space before the equal sign?

jiakai-17 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,358 @@
+            break;
+        }
+
+        FileWriter fw = null;
+
+        try {
+            fw = new FileWriter(dataFile, true);
+            fw.write(task.exportSaveFormat() + System.lineSeparator());
+            tasks.add(task);
+
+            System.out.println(INDENTATION + "Got it, I've added this task:" + System.lineSeparator()
+                    + INDENTATION + task.toString());
+            System.out.printf(INDENTATION + "Now you have %d tasks in the list\n", tasks.size());
+            showSeparationLine();
+        } catch (IOException e) {
+            throw new TasketException("IO Error");
+        } finally {
+            try {
+                if (fw != null) {
+                    fw.close();
+                }
+            } catch (IOException e) {
+                throw new TasketException("IO Error");
+            }
+        }
```
Perhaps you could use Java's try-with-resources to eliminate the extra try-catch in the finally block?

jiakai-17 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,358 @@
+        case "event":
+            task = createEventTask(prompt);
+            break;
+        }
```
Perhaps you could add a default case for the switch statement, allowing you to catch errors.

jiakai-17 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,358 @@
+            System.out.printf(INDENTATION + "Now you have %d tasks in the list\n", tasks.size());
+            showSeparationLine();
+        } catch (NumberFormatException e) {
+            throw new TasketException("The task index must be a number");
```
Perhaps you could extract these strings and replace them with constants?

jiakai-17 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,358 @@
+        String action = prompt.split(" ", 2)[0];
+
+        switch (action) {
+        case "todo":
+        case "deadline":
+        case "event":
```
Perhaps you could add the `//Fallthrough` comment here to explain why the break statement is omitted?

jiakai-17 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,358 @@
+        }
+    }
+
+    public static void markTaskUndone(String text) throws TasketException {
```
`text` param is unused here. Perhaps you meant `prompt`?

jiakai-17 commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,20 @@
+
+    @Override
+    public String toString() {
+        return String.format("[E]%s (from: %s to: %s)", super.toString(),this.from, this.to);
```
',' is not followed by whitespace. (13:73) [WhitespaceAfter]

jiakai-17 reviewed (commented):
Looks good so far. Good job on following the naming conventions. 
I have left some comments about minor style violations and suggestions.

lerxuann commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,358 @@
+
+    public static void identifyTaskType(String type, String prompt) throws TasketException {
+        Task task = null;
+        switch (type) {
+        case "todo":
+            task = createToDoTask(prompt);
+            break;
+
+        case "deadline":
+            task = createDeadlineTask(prompt);
+            break;
+
+        case "event":
```
good job following the coding standards for the switch case!

lerxuann commented on src/main/java/Duke.java:
Overall missing Javadoc comments, though I understand you may not have had the time to add them

Ken-Lai committed:
Extract closely related code to classes

Ken-Lai committed:
Merge branch 'branch-A-MoreOOP'

Ken-Lai committed:
Divide classes into packages

Ken-Lai committed:
Merge branch 'branch-A-Packages'

Ken-Lai committed:
Merge branch 'add-gradle-support'

Ken-Lai committed:
Automate project builds using Gradle

Ken-Lai committed:
Add JUnit test to test the code

Ken-Lai committed:
Package the app as a jar file

Ken-Lai committed:
Add ability to find task with keywords

Ken-Lai committed:
Tweak the code to comply with java coding standard

Ken-Lai committed:
Add JavaDoc comments to the methods and classes

Ken-Lai committed:
Merge branch 'branch-A-CodingStandard'

Ken-Lai committed:
Merge branch 'branch-Level-9'

Ken-Lai committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/tasket/Duke.java
#	src/main/java/tasket/ui/Ui.java

Ken-Lai committed:
Add JavaDoc comment for FindCommand

Ken-Lai committed:
Improve java coding style with checkstyle

Ken-Lai committed:
Merge branch 'branch-A-CheckStyle'

Ken-Lai committed:
Add GUI for tasket.

Ken-Lai committed:
Merge branch 'branch-Level-10'

Ken-Lai committed:
Amend JavaDoc for FindCommand

Ken-Lai committed:
Document important assumptions with assert

The use of assertions is to ensure the assumptions on every step of
code can be true.

If an error occurs during assertions, we can just trace the steps
with a debugger, which makes debugging easier.

Ken-Lai committed:
Refactor the code to improve code quality

The codes are refactored according to 'Code Quality' topics. This
improves the readability of the code.

Ken-Lai committed:
Merge pull request #2 from Ken-Lai/branch-A-CodeQuality

Refactor the code to improve code quality

Ken-Lai committed:
Merge branch 'master' into branch-A-Assertions

# Conflicts:
#	src/main/java/tasket/parser/Parser.java

Ken-Lai committed:
Merge pull request #3 from Ken-Lai/branch-A-Assertions

Document important assumptions with assert

Ken-Lai committed:
Add ability to tag tasks.

Multiple tags can be inserted to a task.

Ken-Lai committed:
Merge branch 'branch-C-Tagging'

Ken-Lai committed:
Add improvements for the gui

Ken-Lai committed:
Merge branch 'branch-A-BetterGui'

Ken-Lai committed:
Add more tests for Tasket.

Ken-Lai committed:
Merge branch 'branch-A-MoreTesting'

Ken-Lai committed:
Add a user guide for tasket.

Ken-Lai committed:
Add changes to UI and README

Ken-Lai committed:
Refactor _config.yml

Ken-Lai committed:
Remove _config.yml and make minor change README.md
